internal/proxy/proxy.go

```diff
@@ -60,38 +61,113 @@ type Client struct {
 	mnts []*socketMount
 }
 
-// NewClient completes the initial setup required to get the proxy to a "steady" state.
-func NewClient(ctx context.Context, cmd *cobra.Command, conf *Config) (*Client, error) {
-	d, err := cloudsqlconn.NewDialer(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("error initializing dialer: %v", err)
+type portConfig struct {
+	mu        sync.Mutex
+	global    int
+	postgres  int
+	mysql     int
+	sqlserver int
+}
+
+func newPortConfig(global int) *portConfig {
+	return &portConfig{
+		global:    global,
+		postgres:  5432,
+		mysql:     3306,
+		sqlserver: 1433,
+	}
+}
+
+// nextPort returns the next port based on the initial global value.
+func (c *portConfig) nextPort() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	p := c.global
+	c.global++
+	return p
+}
+
+func (c *portConfig) nextDBPort(version string) int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	switch {
+	case strings.HasPrefix(version, "MYSQL"):
+		p := c.mysql
+		c.mysql++
+		return p
+	case strings.HasPrefix(version, "POSTGRES"):
+		p := c.postgres
+		c.postgres++
+		return p
+	case strings.HasPrefix(version, "SQLSERVER"):
+		p := c.sqlserver
+		c.sqlserver++
+		return p
+	default:
+		// Unexpected engine version, use global port setting instead.
+		return c.nextPort()
 	}
-	c := &Client{cmd: cmd, dialer: d}
+}
 
-	port := conf.Port
-	var inc int
+// NewClient completes the initial setup required to get the proxy to a "steady" state.
+func NewClient(ctx context.Context, d *cloudsqlconn.Dialer, cmd *cobra.Command, conf *Config) (*Client, error) {
+	pc := newPortConfig(conf.Port)
+	type mountRes struct {
+		mnt *socketMount
+		err error
+	}
+	mntCh := make(chan mountRes, len(conf.Instances))
 	for _, inst := range conf.Instances {
-		m := &socketMount{inst: inst.Name}
-		a := conf.Addr
-		if inst.Addr != "" {
-			a = inst.Addr
-		}
-		nextPort := port + inc
-		if inst.Port != 0 {
-			nextPort = inst.Port
-		} else {
-			inc++
+		go func(i InstanceConnConfig) {
+			m := &socketMount{inst: i.Name}
+			a := conf.Addr
+			if i.Addr != "" {
+				a = i.Addr
+			}
+			version, err := d.EngineVersion(ctx, i.Name)
+			if err != nil {
+				mntCh <- mountRes{err: err}
+				return
+			}
+			var np int
+			switch {
+			case i.Port != 0:
+				np = i.Port
+			case conf.Port != 0:
+				np = pc.nextPort()
+			default:
+				np = pc.nextDBPort(version)
+			}
+			addr, err := m.listen(ctx, "tcp", net.JoinHostPort(a, fmt.Sprint(np)))
+			if err != nil {
+				err := fmt.Errorf("[%v] Unable to mount socket: %v", i.Name, err)
+				mntCh <- mountRes{err: err}
+				return
+			}
+			cmd.Printf("[%s] Listening on %s\n", i.Name, addr.String())
+			mntCh <- mountRes{mnt: m}
+		}(inst)
```

Comment:
This bit concerns me a bit, because I think it makes the port used non-deterministic. This could cause the "auto" port behavior to be inconsistent, which I don't think we want. 

I think what we mostly want to do is run `d.EngineVersion` in parallel - can we just do a quick for loop to "warm up" the call (and disregard the result, we just want it to start the refresh), and then when we call it below it'll block until ready?

Reply:
Sounds good.